common/lib/client/paginatedresource.ts

```diff
@@ -0,0 +1,195 @@
+import * as Utils from '../util/utils';
+import Logger from '../util/logger';
+import Resource from './resource';
+import ErrorInfo from '../types/errorinfo';
+import { PaginatedResultCallback } from '../../types/utils';
+
+// TODO: Replace this once rest.js is converted to TypeScript
+type Rest = any;
+
+export type BodyHandler = (body: unknown, headers: Record<string, string>, packed?: boolean) => any;
+
+function getRelParams(linkUrl: string) {
+	const urlMatch = linkUrl.match(/^\.\/(\w+)\?(.*)$/);
+	return urlMatch && Utils.parseQueryString(urlMatch[2]);
+}
+
+function parseRelLinks(linkHeader: string | Array<string>) {
+	if(typeof(linkHeader) == 'string')
+		linkHeader = linkHeader.split(',');
+
+	const relParams: Record<string, Record<string, string>> = {};
+	for(let i = 0; i < linkHeader.length; i++) {
+		const linkMatch = linkHeader[i].match(/^\s*<(.+)>;\s*rel="(\w+)"$/);
+		if(linkMatch) {
+			const params = getRelParams(linkMatch[1]);
+			if(params)
+				relParams[linkMatch[2]] = params;
+		}
+	}
+	return relParams;
```

Comment:
There are a couple of unsafe array accesses on this function. I am not sure if those functions have their corresponding unit tests but I think it would be a great way to discover and fix those edge cases

Reply:
You're right, but I don't want to change this now. It's always worked up until this point so I would assume it's safe, even if it would take a lot of work to prove it.

Reply:
Resolving this conversation on @ikbalkaya's behalf.